stdlib/importlib/metadata/__init__.pyi

```diff
@@ -35,6 +35,23 @@ if sys.version_info >= (3, 8):
             dist: ClassVar[Distribution | None]
             def matches(self, **params: Any) -> bool: ...  # undocumented
 
+    if sys.version_info >= (3, 10):
+        class EntryPoints(list[EntryPoint]):  # use as list is deprecated since 3.10
+            def select(self, **params: Any) -> EntryPoints: ...
+            @property
+            def names(self) -> set[str]: ...
+            @property
+            def groups(self) -> set[str]: ...
+
+        class SelectableGroups(dict[str, EntryPoints]):  # use as list is deprecated since 3.10
+            @classmethod
+            def load(cls, eps: SelectableGroups) -> SelectableGroups: ...
+            @property
+            def groups(self) -> set[str]: ...
+            @property
+            def names(self) -> set[str]: ...
+            def select(self, **params: Any) -> SelectableGroups: ...
```

Comment:
This should return `EntryPoints` instead.

```suggestion
            def select(self, **params: Any) -> EntryPoints: ...
```

Reply:
Actually it can be either: it returns self if there are no params, but `EntryPoints` if there are any.

Reply:
Didn't realize empty `select()` is possible 😅

Reply:
Not how I would design this API :)

Reply:
Updated, let me know what you think.

Reply:
I believe this needs an overload:

```python
@overload
def select(self: Self) -> Self: ...
@overload
def select(self, **params: Any) -> EntryPoints: ...
```

This might need a `# type: ignore` on the second overload.

Reply:
> This might need a `# type: ignore` on the second overload.

To silence this `mypy` error?
```
stdlib\importlib\metadata\__init__.pyi:54: error: Overloaded function signatures 1 and 2 overlap with incompatible return types
```

`mypy` reports the error on the first overload so comment should be on the first one I think.

Reply:
You are right, first overload it is.

Reply:
By the way, why does mypy report this as an error? Dynamic return types aren't that uncommon.

Reply:
Added the overload.

Reply:
Good question. This needs to be commonly ignored in typeshed. I'd be happy if mypy didn't warn about this.

Reply:
The idea is probably that the return type of `select(**some_dict)` depends on whether the dict is empty or not. So the best that mypy can do is to return a union, which is annoying to deal with in practice.

For non-typeshed code, I think getting a warning about this is a good thing. because methods like this would be less confusing if written as two separate methods, and getting a warning will make you rethink about whether it is a good idea.